### PR TITLE
Feat/sign placements and programmes

### PR DESCRIPTION
**Pull request by Judge40**
Adds a signature to placements and programme memberships with the following structure
```json
{
    "tisId": "315",
    ...
    "status": "CURRENT",
    "signature": {
        "signedAt": "2023-01-11T13:41:39.795528900Z",
        "validUntil": "2023-01-12T13:41:39.795528900Z",
        "hmac": "af03b176aff00fd8bb3bf4723ba6ec895cd2e8c94862f63c0e1c33b576b0e2df"
    }
}
```

### CODE REVIEW

File: src/main/java/uk/nhs/hee/trainee/details/service/SignatureService.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+   * @param dto The DTO to sign.
+   * @throws JsonProcessingException If the DTO could not be read as JSON.
+   */
+  public void signDto(SignedDto dto) throws JsonProcessingException {
+    Duration expireAfter = signatureConfigurationProperties.getExpireAfter(dto);
+    Signature signature = new Signature(expireAfter);
+    dto.setSignature(signature);
+
+    String secretKey = signatureConfigurationProperties.getSecretKey();
+    byte[] dtoBytes = mapper.writeValueAsBytes(dto);
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/hee/trainee/details/service/SignatureService.java:67]: Cool, glad to see the hmac is based on the DTO only once the signature signedAt and validUntil properties are set ✔️

File: src/test/java/uk/nhs/hee/trainee/details/service/SignatureServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    long expiresAfter = signedAt.until(signature.getValidUntil(), ChronoUnit.MINUTES);
+    assertThat("Unexpected valid duration.", expiresAfter, is(1440L));
+
+    String hmac = signature.getHmac();
+    assertThat("Unexpected hmac.", hmac, notNullValue());
+    assertThat("Unexpected hmac length.", hmac.length(), is(64));
+  }
+
+  @Test
+  void shouldNotValidateWhenHmacIncluded() throws JsonProcessingException, InterruptedException {
```
[the remark this answers is not in the document]
Comment (Judge40) [reply src/test/java/uk/nhs/hee/trainee/details/service/SignatureServiceTest.java:97]: `shouldNotValidateWhenHmacIncludedInHmacGeneration` and `shouldValidateWhenHmacExcludedFromHmacGeneration`?

[the remark this answers is not in the document]
Comment (Judge40) [reply src/test/java/uk/nhs/hee/trainee/details/service/SignatureServiceTest.java]: @ReubenRobertsHEE done

**Review by ReubenRobertsHEE — APPROVED** [review]
Nice

**Pull request merged**